pkg/monitor/cluster/certificateexpirationstatuses.go

```diff
@@ -0,0 +1,63 @@
+package cluster
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/Azure/ARO-RP/pkg/operator"
+	"github.com/Azure/ARO-RP/pkg/operator/controllers/genevalogging"
+	"github.com/Azure/ARO-RP/pkg/util/dns"
+)
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+func (mon *Monitor) emitCertificateExpirationStatuses(ctx context.Context) error {
+	// report NotAfter dates for Geneva (always), Ingress, and API (on managed domain) certificates
+	var certs []*x509.Certificate
+
+	mdsdCert, err := mon.getCertificate(ctx, operator.SecretName, operator.Namespace, genevalogging.GenevaCertName)
+	if err != nil {
+		return err
+	}
+	certs = append(certs, mdsdCert)
+
+	if dns.IsManagedDomain(mon.oc.Properties.ClusterProfile.Domain) {
+		infraID := mon.oc.Properties.InfraID
+		for _, secretName := range []string{infraID + "-ingress", infraID + "-apiserver"} {
+			certificate, err := mon.getCertificate(ctx, secretName, operator.Namespace, corev1.TLSCertKey)
+			if err != nil {
+				return err
+			}
+			certs = append(certs, certificate)
+		}
+	}
+
+	for _, cert := range certs {
+		mon.emitGauge("certificate.expirationdate", 1, map[string]string{
+			"subject":        cert.Subject.CommonName,
+			"expirationDate": cert.NotAfter.UTC().Format(time.RFC3339),
+		})
+	}
+	return nil
+}
+
+func (mon *Monitor) getCertificate(ctx context.Context, secretName, secretNamespace, secretKey string) (*x509.Certificate, error) {
+	secret, err := mon.cli.CoreV1().Secrets(secretNamespace).Get(ctx, secretName, metav1.GetOptions{})
+	if err != nil {
+		return &x509.Certificate{}, err
+	}
+
+	certBlock, _ := pem.Decode(secret.Data[secretKey])
+	if certBlock == nil {
+		return &x509.Certificate{}, fmt.Errorf(`certificate "%s" not found on secret "%s"`, secretKey, secretName)
```

Comment:
I'm not sure we can/should be logging certificate data in this way into geneva. Using the name of the certificate should be fine since it's one we manage, but we should remove the `secretKey` from the error message

Reply:
I think there's a misunderstanding due to overloading of the term "key". In this case `secretKey` isn't the TLS key for the certificate; those are never accessed anywhere in this code. It's the "key" as in "key/value pair" of the data on the object, and it comes in through the call to `getCertificate()`. It's either `gcscert.pem` in the case of the Geneva certificate (line 31), or `tls.crt` in the case of the native k8s TLS Secret objects (line 50).